subsys/usb/device_next/usbd_msg.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/usb/usbd.h>
+
+#include "usbd_device.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_msg, CONFIG_USBD_LOG_LEVEL);
+
+static void msg_work_handler(struct k_work *work);
+static K_WORK_DEFINE(msg_work, msg_work_handler);
+K_FIFO_DEFINE(msg_queue);
+
+struct usbd_msg_pkt {
+	sys_snode_t node;
+	struct usbd_contex *ctx;
+	struct usbd_msg msg;
+};
+
+K_MEM_SLAB_DEFINE_STATIC(usbd_msg_slab, sizeof(struct usbd_msg_pkt),
+			 CONFIG_USBD_MSG_SLAB_COUNT, sizeof(void *));
+
+static ALWAYS_INLINE void submit_new_work(struct usbd_msg_pkt *const m_pkt)
+{
+	k_fifo_put(&msg_queue, m_pkt);
+	if (k_work_submit(&msg_work) < 0) {
+		__ASSERT(false, "Failed to submit work");
+	}
+}
+
+static inline struct usbd_msg_pkt *usbd_alloc_msg(struct usbd_contex *const ctx)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (k_mem_slab_alloc(&usbd_msg_slab, (void **)&m_pkt, K_NO_WAIT)) {
+		LOG_ERR("Failed to allocate message memory");
+		return NULL;
+	}
+
+	memset(m_pkt, 0, sizeof(struct usbd_msg_pkt));
+	m_pkt->ctx = ctx;
+
+	return m_pkt;
+}
+
+static void msg_work_handler(struct k_work *work)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	m_pkt = k_fifo_get(&msg_queue, K_NO_WAIT);
+	if (m_pkt != NULL) {
+		m_pkt->ctx->msg_cb(&m_pkt->msg);
+		k_mem_slab_free(&usbd_msg_slab, (void *)m_pkt);
+	}
+
+	if (!k_fifo_is_empty(&msg_queue)) {
+		(void)k_work_submit(work);
+	}
+}
+
+int usbd_msg_register_cb(struct usbd_contex *const uds_ctx,
+			 const usbd_msg_cb_t cb)
+{
+	int ret = 0;
+
+	usbd_device_lock(uds_ctx);
+
+	if (uds_ctx->msg_cb != NULL) {
+		ret = -EALREADY;
+		goto register_cb_exit;
+	}
+
+	uds_ctx->msg_cb = cb;
+
+register_cb_exit:
+	usbd_device_unlock(uds_ctx);
+
+	return ret;
+}
+
+void usbd_msg_pub_simple(struct usbd_contex *const ctx,
+			 const enum usbd_msg_type type, const int status)
+{
+	struct usbd_msg_pkt *m_pkt;
+
+	if (ctx->msg_cb == NULL) {
+		return;
+	}
+
+	m_pkt = usbd_alloc_msg(ctx);
+	if (m_pkt == NULL) {
+		__ASSERT(false, "Failed to allocate message memory");
```

Comment:
There is no other channel to report this, the alternative is LOG_ERR or k_panic() or to block in alloc(). @tmon-nordic You preferences?

Reply:
LOG_ERR doesn't make sense here because it will flood the log (if we had equivalent to Linux log once methods then maybe...). k_panic is similar to assert here and therefore not proper. Blocking in alloc is better than other alternatives. There's also possibility to simply not log it at all and not block.

Reply:
LOG_ERR doesn't make sense here because it will flood the log (if we had equivalent to Linux log once methods then maybe...). k_panic is similar to assert here and therefore not proper. Blocking in alloc is better than other alternatives. There's also possibility to simply not log it at all and not block.

Reply:
> There's also possibility to simply not log it at all and not block.

Would it not be equivalent to assert abuse? assert is also used in the tree for troubleshooting.

Reply:
Here only tracing is appropriate. Assert is used for troubleshooting because it is designed to expose incorrect code, e.g. calling function with invalid parameters and ao on. Abusing asserts for normal runtime behavior is wrong.

Reply:
Here only tracing is appropriate. Assert is used for troubleshooting because it is designed to expose incorrect code, e.g. calling function with invalid parameters and ao on. Abusing asserts for normal runtime behavior is wrong.

Reply:
@tmon-nordic Please recheck.